frame/society/src/migrations.rs

```diff
@@ -0,0 +1,296 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Migrations for Society Pallet
+
+use super::*;
+use frame_support::traits::{Instance, OnRuntimeUpgrade};
+
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+
+/// The log target.
+const TARGET: &'static str = "runtime::society::migration";
+
+/// This migration moves all the state to v2 of Society.
+pub struct MigrateToV2<T: Config<I>, I: 'static, PastPayouts>(
+	sp_std::marker::PhantomData<(T, I, PastPayouts)>,
+);
+
+impl<
+		T: Config<I>,
+		I: Instance + 'static,
+		PastPayouts: Get<Vec<(<T as frame_system::Config>::AccountId, BalanceOf<T, I>)>>,
+	> OnRuntimeUpgrade for MigrateToV2<T, I, PastPayouts>
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+		assert!(
+			can_migrate::<T, I>(),
+			"Invalid (perhaps already-migrated) state detected prior to migration"
+		);
+		Ok(Vec::new())
+	}
```

Comment:
Since we now leave migrations hanging about in the runtime for a few versions after they are applied, pre/post hooks should not start failing post-upgrade. This should be adjusted to log a warning rather than assert.